src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java

```diff
@@ -1,74 +1,167 @@
-
-            SignalConsumerComponent signalConsumerComponent = entity.getComponent(SignalConsumerComponent.class);
-            RailSwitchComponent railSwitchComponent = entity.getComponent(RailSwitchComponent.class);
-
-            if (block == railSwitchSignalOff && consumerStatusComponent.hasSignal) {
-                railSwitchSignalOn.setKeepActive(true);
-                worldProvider.setBlock(blockLocation, railSwitchSignalOn);
-                blockEntityRegistry.getBlockEntityAt(blockLocation).addOrSaveComponent(signalConsumerComponent);
-                if (!blockEntityRegistry.getBlockEntityAt(blockLocation).hasComponent(RailSwitchComponent.class)) {
-                    blockEntityRegistry.getBlockEntityAt(blockLocation).addOrSaveComponent(railSwitchComponent);
+    @ReceiveEvent(components = {BlockComponent.class, RailSwitchComponent.class})
+    public void activateAction(ActivateEvent event, EntityRef entity) {
+        RailSwitchComponent railSwitchComponent = entity.getComponent(RailSwitchComponent.class);
+
+        updateSwitch(event, entity, railSwitchComponent, !railSwitchComponent.isOn, ActionSource.ACTIVATE);
+    }
+
+    @ReceiveEvent(components = {BlockComponent.class, RailSwitchComponent.class, SignalConsumerStatusComponent.class})
+    public void signalAction(OnChangedComponent event, EntityRef entity) {
+        SignalConsumerStatusComponent consumerStatusComponent = entity.getComponent(SignalConsumerStatusComponent.class);
+        RailSwitchComponent railSwitchComponent = entity.getComponent(RailSwitchComponent.class);
+
+        if (consumerStatusComponent.hasSignal != railSwitchComponent.hasSignal) {
+            updateSwitch(event, entity, railSwitchComponent, consumerStatusComponent.hasSignal, ActionSource.SIGNAL);
+        }
+    }
+
+    private void updateSwitch(Event event, EntityRef entity, RailSwitchComponent oldRailComponent, boolean input, ActionSource actionSource) {
+        Vector3i blockLocation = new Vector3i(entity.getComponent(BlockComponent.class).getPosition());
+        switch (oldRailComponent.mode) {
+            case 0:
+                if (input) {
+                    railSwitchLeverOn.setKeepActive(true);
+                    worldProvider.setBlock(blockLocation, railSwitchLeverOn);
+                } else {
+                    railSwitchLeverOff.setKeepActive(true);
+                    worldProvider.setBlock(blockLocation, railSwitchLeverOff);
+                }
+                break;
+            case 1:
+                if (input) {
+                    railSwitchSignalOn.setKeepActive(true);
+                    worldProvider.setBlock(blockLocation, railSwitchSignalOn);
+                } else {
+                    railSwitchSignalOff.setKeepActive(true);
+                    worldProvider.setBlock(blockLocation, railSwitchSignalOff);
+                }
+                break;
+
+            case 2: //TODO: railSwitch using both activate (E button) and Signalling mode
```

Comment:
If this is new code you may want to leave a TODO but not a ton of commented out code. Either reference somewhere for inspiration (copy pasting) in the future or add the details to an issue in the repo instead :-)